File: modulesCore/homeScreen/src/main/java/cn/pivotstudio/modulec/homescreen/network/HSRequestInterface.java
package cn.pivotstudio.modulec.homescreen.network;

import java.util.List;

import cn.pivotstudio.modulec.homescreen.model.DetailForestHole;
import cn.pivotstudio.modulec.homescreen.model.ForestCard;
import cn.pivotstudio.modulec.homescreen.model.ForestCardList;
import cn.pivotstudio.modulec.homescreen.model.ForestHead;
import cn.pivotstudio.modulec.homescreen.model.ForestHeads;
import cn.pivotstudio.modulec.homescreen.model.ForestHole;
import cn.pivotstudio.modulec.homescreen.model.ForestTypes;
import io.reactivex.Observable;
import retrofit2.Response;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.HTTP;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;
import retrofit2.http.Url;

/**
 * @classname HSRequestInterface
 * @description: 首页网络请求接口
 * @date 2022/5/3 0:42
 * @version:1.0
 * @author: lzt
 */
public interface HSRequestInterface {
    @GET("version")
    Observable<VersionResponse> checkUpdate();

    @GET("holes?")
    Observable<List<HomepageHoleResponse.DataBean>> homepageHoles(@Query("is_descend") Boolean is_descend,
                                                                  @Query("is_last_reply") Boolean is_last_reply,
                                                                  @Query("start_id") int start_id,
                                                                  @Query("list_size") int list_size);

    @POST
    Observable<MsgResponse> follow(@Url String url);

    @DELETE
    Observable<MsgResponse> deleteFollow(@Url String url);

    @POST
    Observable<MsgResponse> report(@Url String url);

    @POST
    Observable<MsgResponse> thumbups(@Url String url);

    @DELETE
    Observable<MsgResponse> deleteThumbups(@Url String url);

    @HTTP(method = "DELETE", path = "holes/{hole_id}", hasBody = false)
    Observable<MsgResponse> deleteHole(@Path("hole_id") String hole_id);

    @GET
    Observable<HomepageHoleResponse.DataBean> searchSingleHole(@Url String url);

    @GET("search/hole?")
    Observable<List<HomepageHoleResponse.DataBean>> searchHoles(@Query("keyword") String keyword,
                                                                @Query("start_id") int start_id,
                                                                @Query("list_size") int list_size);

    @GET("forests/holes?")
    Observable<List<ForestHole>> searchForestHoles(
            @Query("start_id") int startId,
            @Query("list_size") int listSize,
            @Query("is_last_reply") Boolean isLastReply
    );

    @GET("forests/joined?")
    Observable<ForestHeads> searchForestHeads(
            @Query("start_id") int startId,
            @Query("list_size") int listSize
    );

    @GET("forests/types?")
    Observable<ForestTypes> searchForestTypes(
            @Query("start_id") int startId,
            @Query("list_size") int listSize
    );

    @GET("forests/type/{forest_type}")
    Observable<ForestCardList> searchForestByType(
            @Path("forest_type") String type,
            @Query("start_id") int startId,
            @Query("list_size") int listSize
    );

    @GET("forests/{forest_id}/holes")
    Observable<List<DetailForestHole>> searchDetailForestHolesByForestId(
            @Path("forest_id") int forestId,
            @Query("start_id") int startId,
            @Query("list_size") int listSize
    );

    @GET("forests/detail/{forest_id}")
    Observable<ForestCardList> searchDetailForestOverviewByForestId(
            @Path("forest_id") int forestId
    );
}

File: modulesCore/homeScreen/src/main/java/cn/pivotstudio/modulec/homescreen/network/HSRequestInterface.kt
package cn.pivotstudio.modulec.homescreen.network

import cn.pivotstudio.husthole.moduleb.network.model.DetailForestHole
import cn.pivotstudio.husthole.moduleb.network.model.ForestHole
import cn.pivotstudio.modulec.homescreen.model.*
import cn.pivotstudio.modulec.homescreen.network.HomepageHoleResponse.DataBean
import io.reactivex.Observable
import retrofit2.http.*

/**
 * @classname HSRequestInterface
 * @description: 首页网络请求接口
 * @date 2022/5/3 0:42
 * @version:1.0
 * @author: mhh
 */
interface HSRequestInterface {
    @GET("version")
    fun checkUpdate(): Observable<VersionResponse>

    @GET("holes")
    fun homepageHoles(
        @Query("is_descend") is_descend: Boolean,
        @Query("is_last_reply") is_last_reply: Boolean,
        @Query("start_id") start_id: Int,
        @Query("list_size") list_size: Int
    ): Observable<List<DataBean>>

    @POST
    fun follow(@Url url: String): Observable<MsgResponse>

    @DELETE
    fun deleteFollow(@Url url: String): Observable<MsgResponse>

    @POST
    fun thumbups(@Url url: String): Observable<MsgResponse>

    @DELETE
    fun deleteThumbups(@Url url: String): Observable<MsgResponse>

    @HTTP(method = "DELETE", path = "holes/{hole_id}", hasBody = false)
    fun deleteHole(@Path("hole_id") holeId: String): Observable<MsgResponse>

    @GET
    fun searchSingleHole(@Url url: String): Observable<DataBean>

    @GET("search/hole")
    fun searchHoles(
        @Query("keyword") keyword: String,
        @Query("start_id") start_id: Int,
        @Query("list_size") list_size: Int
    ): Observable<List<DataBean>>

    @GET("notices")
    fun searchNotices(
        @Query("start_id") startId: Int,
        @Query("list_size") listSize: Int
    ): Observable<NoticeResponse>

    @GET("notices")
    suspend fun searchNoticesFlow(
        @Query("start_id") startId: Int,
        @Query("list_size") listSize: Int
    ): NoticeResponse


    /** =========小树林相关的接口========== */
    @GET("forests/holes")
    fun searchForestHoles(
        @Query("start_id") startId: Int,
        @Query("list_size") listSize: Int,
        @Query("is_last_reply") isLastReply: Boolean
    ): Observable<List<ForestHole>>

    @GET("forests/joined")
    fun searchForestHeads(
        @Query("start_id") startId: Int,
        @Query("list_size") listSize: Int
    ): Observable<ForestHeads>

    @POST("forests/join/{forest_id}")
    fun joinTheForest(
        @Path("forest_id") forestId: String
    ): Observable<MsgResponse>

    @DELETE("forests/quit/{forest_id}")
    fun quitTheForest(
        @Path("forest_id") forestId: String
    ): Observable<MsgResponse>

    @GET("forests/types")
    fun searchForestTypes(
        @Query("start_id") startId: Int,
        @Query("list_size") listSize: Int
    ): Observable<ForestTypes>

    @GET("forests/type/{forest_type}")
    fun searchForestByType(
        @Path("forest_type") type: String,
        @Query("start_id") startId: Int,
        @Query("list_size") listSize: Int
    ): Observable<ForestCardList>

    @GET("forests/{forest_id}/holes")
    fun searchDetailForestHolesByForestId(
        @Path("forest_id") forestId: Int,
        @Query("start_id") startId: Int,
        @Query("list_size") listSize: Int
    ): Observable<List<DetailForestHole>>

    @GET("forests/detail/{forest_id}")
    fun searchDetailForestOverviewByForestId(
        @Path("forest_id") forestId: Int
    ): Observable<ForestCardList>
}